src/librustc_middle/traits/mod.rs

```diff
@@ -80,8 +81,33 @@ pub enum Reveal {
 }
 
 /// The reason why we incurred this obligation; used for error reporting.
+///
+/// As the happy path does not care about this struct, we store it on the heap
+/// to improve the performance of the happy path.
+///
+/// We do not want to intern this as there are a lot of obligation causes which
+/// only live for a short period of time.
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub struct ObligationCause<'tcx> {
+    data: Lrc<ObligationCauseData<'tcx>>,
+}
+
+impl Deref for ObligationCause<'tcx> {
+    type Target = ObligationCauseData<'tcx>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.data
+    }
+}
+
+impl DerefMut for ObligationCause<'tcx> {
```

Comment:
Can we get rid of this? I'm suspicious of code that is mutating an `ObligationCause` in-place. If some exists that is not trivially refactorable, I think you should add an inherent `make_mut` method instead so the cost of doing this becomes clear.